### PR TITLE
support Artifactory NuGet V3 API

### LINKED ISSUE #579: Artifactory issues vi APIv3
It seems Artifactory is not following the standards completely and so NugetForUnity does not work with it at the moment.

1. The first issue is that NugetForUnity only recognized a Source as V3 if it ends with `index.json` but Artifactory nuget url looks like this: `https://artifactory.in.mycompany.com/artifactory/api/nuget/v3/mycompany-nuget-local`. We need to also recognize source as V3 if its url contains '/v3/'. We should also add the '/' at the end if it is not entered since currently it doesn't work unless slash is manually added.
2. Second issue is that NugetApiClientV3.InitializeApiAddresses method will not find `PackageBaseAddress/3.0.0` resource type which will log the `The NuGet package source at '{0}' has no PackageBaseAddress resource defined.` error and package download will not work. I will attach the Artifactory [response](https://github.com/GlitchEnzo/NuGetForUnity/files/12837869/mycompany-nuget-local.zip) in case someone figures out a better way but it seems none of the links there specify a proper download link which looks something like this: `https://artifactory.in.mycompany.com:443/artifactory/api/nuget/mycompany-nuget-local/Download/mycompany.litjson/1.0.87`. I got to this URL by first finding `@id` field of the specific package version in query response to be `https://artifactory.in.mycompany.com:443/artifactory/api/nuget/v3/mycompany-nuget-local/registration/mycompany.litjson/1.0.87.json` and then after opening that url the `packageContent` field had the download URL. It seems that same fields can be followed on main nuget repository.
3. Third issue is that artifactory doesn't support queries in form of `packageId:ID1 packageId:ID2`, and I can't figure out how to make a query like that work. I thought I saw it worked when I used comma instead but it seems the response was not right. Comma does not work with main nuget repo anyway so we have to figure out what to do here. The change should be made in NugetPackageSourceV3.GetUpdates.

### PR DESCRIPTION
**Pull request by JoC0de**
fixes #579 
* fix preference screen switching to different nuget source type
* allow using authenticated V3 api
* allow overwriting nuget source api version
* allow using encrypted nuget source passwords
* fix some warnings

To fix the issue that Artifactory doesn't have a `PackageBaseAddress` I used a dialog:
image
If clicked on "Configure for Artifactory" the configuration is updated to the following:
image

So I added `PackageDownloadUrlTemplateOverwrite` witch should allow even customization for other non standard NuGet V3 API's

### CODE REVIEW

**Review by popara96 — APPROVED** [review]

File: src/NuGetForUnity/Editor/PackageSource/NugetApiClientV3.cs [hunk 1]
```diff
@@ -414,26 +463,62 @@ private async void InitializeApiAddresses(NugetPackageSourceV3 packageSource)
...
+                                "Missing 'PackageBaseAddress' endpoint",
+                                $"The used NuGet V3 API '{apiIndexJsonUrl}' doesn't support the 'PackageBaseAddress' endpoint. You need to provide the url manually by specifying '{NugetConfigFile.PackageDownloadUrlTemplateOverwriteAttributeName}' on the package source. If this is a NuGet source provided by Artifactory you can click on 'Configure for Artifactory' to configure it.",
+                                "Configure for Artifactory",
+                                "Cancel");
+                            if (displayDialog)
+                            {
+                                packageDownloadUrlTemplate = $"{registrationsBaseUrl}Download/{{0}}/{{1}}";
+                                PackageDownloadUrlTemplateOverwrite = packageDownloadUrlTemplate;
+                                packageSource.UpdateSearchBatchSize =
+                                    1; // Artifactory somehow can't handle search queries containing multiple packageId's.
```
Comment (igor84) [inline src/NuGetForUnity/Editor/PackageSource/NugetApiClientV3.cs:485]: I was hoping you will find some magic incantation to make it work :). I'll see if I can open an issue to Artifactory about this tomorrow.

**Review by igor84 — APPROVED** [review]

File: src/NuGetForUnity.Cli/Fakes/Mathf.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+using System.Diagnostics.CodeAnalysis;
+
+namespace UnityEngine
+{
+    [SuppressMessage("ReSharper", "IdentifierTypo", Justification = "Conform with Unity naming.")]
+    internal static class Mathf
+    {
+        public static int Clamp(int value, int min, int max)
+        {
+            return Math.Clamp(value, min, max);
```
[the remark this answers is not in the document]
Comment (JoC0de) [reply src/NuGetForUnity.Cli/Fakes/Mathf.cs:13]: I got (from the CI):
```
/github/workspace/src/NuGetForUnity/Editor/Configuration/NugetConfigFile.cs(180,67): error CS0117: 'Math' does not contain a definition for 'Clamp'
```
I was also really confused but Clamp was added in .net standard 2.1 (again a reason to update Unity)

File: src/NuGetForUnity/Editor/PackageSource/NugetApiClientV3.cs [hunk 3]
```diff
@@ -375,8 +391,41 @@ private string GetSessionStateKey()
...
+
+            initializationTaskCompletionSource = new TaskCompletionSource<bool>();
+            _ = InitializeApiAddresses(packageSource);
+
+            return initializationTaskCompletionSource.Task;
+        }
+
+        private async Task InitializeApiAddresses(NugetPackageSourceV3 packageSource)
+        {
+            Debug.Assert(initializationTaskCompletionSource != null, "initializationTaskCompletionSource != null");
```
[the remark this answers is not in the document]
Comment (JoC0de) [reply src/NuGetForUnity/Editor/PackageSource/NugetApiClientV3.cs:428]: I can parse it to skip the nullable check, but I still need it to be a field. The whole reason of the `initializationTaskCompletionSource` is to support multiple waiting tasks. A normal `Task` can only be awaited by a single caller, a `initializationTaskCompletionSource.Task` can be reused -> multiple waiters.

File: src/NuGetForUnity.Cli/Fakes/EditorUtility.cs [hunk 4]
```diff
@@ -14,6 +14,29 @@ internal static void ClearProgressBar()
             currentProgressBarTitle = string.Empty;
         }
 
+        internal static bool DisplayDialog(string title, string message, string ok, string cancel)
```
[the remark this answers is not in the document]
Comment (JoC0de) [reply src/NuGetForUnity.Cli/Fakes/EditorUtility.cs:17]: I also wasn't sure if I should either

1. throw a expectation
2. always ask for input (probably the cli get stuck waiting for user input)
3. add a `--non-interactive` command argument to disable console input

but yes as the CLI can't be used for package installation / only for package restore the probability of getting into this error and calling it in a interactive console is very low.

[the remark this answers is not in the document]
Comment (JoC0de) [reply src/NuGetForUnity.Cli/Fakes/EditorUtility.cs:17]: I changed it to "throw a exception" for now the best solution I think.

File: src/NuGetForUnity/Editor/PackageSource/NugetApiClientV3.cs [hunk 5]
```diff
@@ -414,26 +463,62 @@ private async void InitializeApiAddresses(NugetPackageSourceV3 packageSource)
...
+                        () =>
+                        {
+                            var displayDialog = EditorUtility.DisplayDialog(
+                                "Missing 'PackageBaseAddress' endpoint",
+                                $"The used NuGet V3 API '{apiIndexJsonUrl}' doesn't support the 'PackageBaseAddress' endpoint. You need to provide the url manually by specifying '{NugetConfigFile.PackageDownloadUrlTemplateOverwriteAttributeName}' on the package source. If this is a NuGet source provided by Artifactory you can click on 'Configure for Artifactory' to configure it.",
+                                "Configure for Artifactory",
+                                "Cancel");
+                            if (displayDialog)
+                            {
+                                packageDownloadUrlTemplate = $"{registrationsBaseUrl}Download/{{0}}/{{1}}";
```
[the remark this answers is not in the document]
Comment (JoC0de) [reply src/NuGetForUnity/Editor/PackageSource/NugetApiClientV3.cs:482]: I also tested it against a Artifactory instance also with the encrypted password.

File: src/NuGetForUnity/Editor/Configuration/ConfigurationEncryptionHelper.cs [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+            if (Application.platform != RuntimePlatform.WindowsEditor)
+            {
+                Debug.LogError("Encrypted passwords are not supported on non-Windows platforms.");
+                return value;
+            }
+
+            var decryptedByteArray = Encoding.UTF8.GetBytes(value);
+
+#if UNITY_EDITOR_WIN
+            var encryptedByteArray = ProtectedData.Protect(decryptedByteArray, EntropyBytes, DataProtectionScope.CurrentUser);
```
[the remark this answers is not in the document]
Comment (JoC0de) [reply src/NuGetForUnity/Editor/Configuration/ConfigurationEncryptionHelper.cs:46]: I got it working for 2023.1.11f somehow Unity ships the Editor with the `System.Security.Cryptography.ProtectedData.dll`, placed inside `C:\Program Files\Unity\Hub\Editor\2023.1.1f1\Editor\Data\NetStandard\EditorExtensions\System.Security.Cryptography.ProtectedData.dll`. With this it even works wen targeting .netstandard2.1.
I couldn't find any information about when Unity added this file to the Editor. For now I changed the pragma to include `UNITY_2023_1_OR_NEWER`. I will try to find some time to test it on 2022.

File: src/NuGetForUnity/Editor/PackageSource/NugetApiClientV3.cs [hunk 7]
```diff
@@ -375,8 +391,41 @@ private string GetSessionStateKey()
...
+                return successful;
+            }
+            catch (Exception exception)
+            {
+                Debug.LogError($"Initialization of api client for '{apiIndexJsonUrl}' failed so we can't use it. error (cached): {exception}.");
+                return false;
+            }
+        }
+
+        private Task<bool> AwaitableInitializeApiAddresses(NugetPackageSourceV3 packageSource)
```
[the remark this answers is not in the document]
Comment (JoC0de) [reply src/NuGetForUnity/Editor/PackageSource/NugetApiClientV3.cs]: I run a Analyzer and fixed all violations. I always forgot this without having the warning.

**Pull request merged**